iroh/src/discovery.rs

```diff
@@ -147,14 +148,14 @@ pub mod static_provider;
 ///
 /// [`RelayUrl`]: crate::relay::RelayUrl
 pub trait Discovery: std::fmt::Debug + Send + Sync {
-    /// Publishes the given [`AddrInfo`] to the discovery mechanism.
+    /// Publishes the given [`RelayUrl`] and direct addreesses to the discovery mechanism.
     ///
     /// This is fire and forget, since the [`Endpoint`] can not wait for successful
     /// publishing. If publishing is async, the implementation should start it's own task.
     ///
     /// This will be called from a tokio task, so it is safe to spawn new tasks.
     /// These tasks will be run on the runtime of the [`super::Endpoint`].
-    fn publish(&self, _info: &AddrInfo) {}
+    fn publish(&self, _url: Option<&RelayUrl>, _addrs: &BTreeSet<SocketAddr>) {}
```

Comment:
Can we keep something like AddrInfo just for discovery? I agree that having AddrInfo in the main API was a bit unpleasant, but I think this change here is not an improvement.

Reply:
In the context of discovery these two things (direct addresses and relay url) are **always** used together. In my opinion it is a very clear case for having a struct. If this PR was new my review comment would be that it's good code but it would be nice to group the arguments for publish in a struct. 🤷‍♂️

Reply:
Not sure I agree, I agonised and stalled  over this when I wrote my version of this a while back.  But now that DiscoveryItem can just contain NodeAddr (NodeId used to not be part of it before) this option of having two args for publish is very reasonable.

The changes to the discovery services also don't indicate AddrInfo is being replicated.

Reply:
I think this is a better api than a struct with only two fields, if we were expecting more arguments, that might be different, but I think having the two explicitly in there is actually good

Reply:
🤷‍♂️ I still don't like it very much but am not extremely opposed to this

 However, didn't we have the idea that publishing could in the future contain an optional ALPN in addition to relay url and direct addresses? At least we discussed this during the retreat.

Reply:
I expect the API to change in some form, but I guess that needs some more design work in general